modules/sync/lock.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package sync
+
+import (
+	"sync"
+
+	"code.gitea.io/gitea/modules/nosql"
+	"code.gitea.io/gitea/modules/setting"
+	redsync "github.com/go-redsync/redsync/v4"
+	goredis "github.com/go-redsync/redsync/v4/redis/goredis/v8"
+)
+
+type Locker interface {
+	Lock() error
+	Unlock() (bool, error)
+}
+
+type LockService interface {
+	GetLock(name string) Locker
+}
+
+type memoryLock struct {
+	mutex *sync.Mutex
+}
+
+func (r *memoryLock) Lock() error {
+	r.mutex.Lock()
+	return nil
+}
+
+func (r *memoryLock) Unlock() (bool, error) {
+	r.mutex.Unlock()
+	return true, nil
+}
+
+var _ Locker = &memoryLock{}
+
+type memoryLockService struct {
+	lockes sync.Map
+}
+
+var _ LockService = &memoryLockService{}
+
+func NewMemoryLockService() *memoryLockService {
+	return &memoryLockService{}
+}
+
+func (l *memoryLockService) GetLock(name string) Locker {
+	lock, _ := l.lockes.LoadOrStore(name, &sync.Mutex{})
```

Comment:
This will add more and more items to the map but never release them.

Reply:
@lunny The closer may work but we could use the locker code from the moby project. It does exactly what we need and has some additional logic we may need but don't know about yet.

https://github.com/moby/locker

Reply:
IMO the closer doesn't work properly.

Think about a case:

* thread-1: mutex.Unlock() 
* thread-2: get the Locker from map
* thread 1: delete Locker from map
* thread 3: get the Locker from map (a new locker, different from the one hold by thread-2)

Reply:
That may be the reason for the waiting counter in locker code from moby.

Reply:
Yup .... multiple-thread problem is always more complex than it looks ....

Reply:
@KN4CK3R @wxiaoguang resolved.